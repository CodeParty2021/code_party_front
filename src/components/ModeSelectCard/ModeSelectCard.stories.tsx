import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { ModeSelectCard } from "./ModeSelectCard";
import { Description } from "pages/Code/Coding/components/Description/Description";

export default {
  title: "components/ModeSelectCard",
  component: ModeSelectCard,
} as ComponentMeta<typeof ModeSelectCard>;

const Template: ComponentStory<typeof ModeSelectCard> = (args) => (
  <ModeSelectCard {...args} />
);

export const Default = Template.bind({});
Default.args = {
  imageName: "grab_flag_lobo",
  title: "タイトルテキスト",
  description: <Description>中身の説明文</Description>,
  icon: "setting",
  disabled: false,
};

export const Disabled = Template.bind({});
Disabled.args = {
  imageName: "grab_flag_lobo",
  title: "タイトルテキスト",
  description: <Description>中身の説明文</Description>,
  icon: "setting",
  disabled: true,
};
