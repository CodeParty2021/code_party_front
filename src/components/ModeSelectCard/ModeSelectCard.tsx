import { Setting, Sword } from "components/icons";
import React, { ReactNode } from "react";
import {
  ModeSelectCardConvexStyle,
  ModeSelectCardStyle,
  ModeSelectCardStyleProps,
} from "./ModeSelectCardStyle";

type Props = ModeSelectCardStyleProps & {
  imageName: "grab_flag_lobo" | "walking_lobo";
  title: string;
  description: ReactNode;
  icon: "setting" | "sword";
  disabled?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
};

type DescriptionTextProps = {
  children: ReactNode;
};
export const DescriptionText: React.FC<DescriptionTextProps> = ({
  children,
}) => {
  return <> {children} </>;
};

export const ModeSelectCard: React.FC<Props> = ({
  imageName,
  title,
  description,
  icon,
  disabled,
  onClick,
  ...styleProps
}) => {
  const Icon = icon === "setting" ? Setting : Sword;
  return (
    <ModeSelectCardStyle disabled={disabled} onClick={onClick} {...styleProps}>
      <div className="modeselectcard_container_image">
        <img src={`/img/modeselectcard_${imageName}.png`} alt={title} />
        <ModeSelectCardConvexStyle
          src="/img/modeselectcard_convex.svg"
          wrapper="svg"
          disabled={disabled}
          {...styleProps}
        />
        {disabled && (
          <div className="modeselectcard_container_info">
            <span>現在は利用できません</span>
          </div>
        )}
      </div>
      <div className="modeselectcard_container_main">
        <Icon size={64} />
        <span className="modeselectcard_modename">{title}</span>
        <span className="modeselectcard_description">{description}</span>
      </div>
      <div className="modeselectcard_border" />
    </ModeSelectCardStyle>
  );
};
