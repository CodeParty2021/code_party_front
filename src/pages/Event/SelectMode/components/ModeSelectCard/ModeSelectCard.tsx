import { Setting, Sword } from "components/icons";
import React from "react";
import {
  ModeSelectCardConvexStyle,
  ModeSelectCardStyle,
  ModeSelectCardStyleProps,
} from "./ModeSelectCardStyle";

type Props = ModeSelectCardStyleProps & {
  mode: "solo" | "battle";
  disabled?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
};

export const ModeSelectCard: React.FC<Props> = ({
  mode,
  disabled,
  onClick,
  ...styleProps
}) => {
  return (
    <ModeSelectCardStyle disabled={disabled} onClick={onClick} {...styleProps}>
      <div className="modeselectcard_container_image">
        {mode == "solo" && (
          <img src="/img/modeselectcard_solo.png" alt="訓練モード" />
        )}
        {mode == "battle" && (
          <img src="/img/modeselectcard_battle.png" alt="対戦モード" />
        )}
        <ModeSelectCardConvexStyle
          src="/img/modeselectcard_convex.svg"
          wrapper="svg"
          disabled={disabled}
          {...styleProps}
        />
        {disabled && (
          <div className="modeselectcard_container_info">
            <span>対戦できるロボットがありません</span>
          </div>
        )}
      </div>
      <div className="modeselectcard_container_main">
        {mode == "solo" && (
          <>
            <Setting size={64} />
            <span className="modeselectcard_modename">訓練モード</span>
            <span className="modeselectcard_description">
              まずはこのモードでロボットを
              <br />
              動かす練習をしましょう。
            </span>
          </>
        )}
        {mode == "battle" && (
          <>
            <Sword size={64} />
            <span className="modeselectcard_modename">対戦モード</span>
            <span className="modeselectcard_description">
              AIが搭載されたコンピュータと
              <br />
              対戦することが出来ます。
            </span>
          </>
        )}
      </div>
      <div className="modeselectcard_border" />
    </ModeSelectCardStyle>
  );
};
