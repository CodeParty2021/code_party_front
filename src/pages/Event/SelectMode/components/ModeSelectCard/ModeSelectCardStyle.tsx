import { ReactSVG } from "react-svg";
import styled, { css } from "styled-components";
import { modeSelectCard } from "styles/colors";
import { FONT_WEIGHT } from "styles/constants/constants";
import { FlexGap } from "styles/FlexGap/FlexGap";

export type ModeSelectCardStyleProps = {};

type localStyleProps = {
  disabled?: boolean;
};

const defaultStyle = css`
  display: flex;
  flex-direction: column;
  overflow: hidden;

  width: 616px;
  height: 744px;
  position: relative;

  background: ${modeSelectCard.bg};
  border-radius: 64px;

  font-weight: ${FONT_WEIGHT.BOLD};
  color: ${modeSelectCard.font};
  fill: ${modeSelectCard.font};

  .modeselectcard_container_image {
    height: 412px;
    align-self: stretch;
    position: relative;

    & > img {
      height: 100%;
      width: 100%;
      object-fit: cover;
    }
  }

  .modeselectcard_container_main {
    display: flex;
    flex-direction: column;
    align-items: center;
    ${FlexGap({ gap: "32px", direction: "column" })}
    padding: 0 16px;
    align-self: stretch;
  }

  .modeselectcard_modename {
    font-size: 30px;
    transform: matrix(1, 0, -0.08, 1, 0, 0);
  }

  .modeselectcard_description {
    font-size: 24px;
    transform: matrix(1, 0, -0.08, 1, 0, 0);
    text-align: start;
  }

  .modeselectcard_border {
    position: absolute;
    width: 100%;
    height: 100%;

    border-radius: 64px;
    box-sizing: border-box;
    border: 2px solid ${modeSelectCard.border};
  }
`;

const hoverStyle = css`
  :hover {
    background: ${modeSelectCard.hover.bg};
    color: ${modeSelectCard.hover.font};
    fill: ${modeSelectCard.hover.font};

    .modeselectcard_border {
      border: 8px solid ${modeSelectCard.hover.border};
    }
  }
`;

const disabledStyle = css`
  .modeselectcard_container_info {
    display: flex;
    ${FlexGap({ gap: "10px", direction: "row" })}
    padding: 16px 24px;

    background: ${modeSelectCard.hoverInfo.bg};
    color: ${modeSelectCard.hoverInfo.font};
    border-radius: 0px 0px 16px 16px;

    position: absolute;
    top: 0;
    left: 50%;
    transform: translate(-50%, -100%);
    transition: transform 0.5s ease;
  }
`;

const disabledHoverStyle = css`
  :hover .modeselectcard_container_info {
    transform: translate(-50%, 0);
  }

  .modeselectcard_container_image > img,
  .modeselectcard_container_main > * {
    opacity: 0.3;
  }
`;

export const ModeSelectCardStyle = styled.button<
  localStyleProps & ModeSelectCardStyleProps
>`
  ${defaultStyle}

  ${({ disabled }) => !disabled && hoverStyle}
  ${({ disabled }) => disabled && disabledStyle}
  ${({ disabled }) => disabled && disabledHoverStyle}
`;

ModeSelectCardStyle.defaultProps = {};

const convexHoverStyle = css`
  ${ModeSelectCardStyle}:hover & {
    fill: ${modeSelectCard.hover.bg};
  }
`;

export const ModeSelectCardConvexStyle = styled(ReactSVG)<
  localStyleProps & ModeSelectCardStyleProps
>`
  width: 184px;
  height: 53px;

  position: absolute;
  left: 50%;
  bottom: 0;
  transform: translateX(-50%);
  fill: ${modeSelectCard.bg};

  ${({ disabled }) => !disabled && convexHoverStyle}
`;
