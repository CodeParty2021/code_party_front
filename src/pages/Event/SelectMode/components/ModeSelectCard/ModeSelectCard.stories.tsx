import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { ModeSelectCard } from "./ModeSelectCard";

export default {
  title: "components/ModeSelectCard",
  component: ModeSelectCard,
} as ComponentMeta<typeof ModeSelectCard>;

const Template: ComponentStory<typeof ModeSelectCard> = (args) => (
  <ModeSelectCard {...args} />
);

export const Default = Template.bind({});
Default.args = {
  mode: "solo",
  disabled: false,
};

export const Disabled = Template.bind({});
Disabled.args = {
  mode: "battle",
  disabled: true,
};
