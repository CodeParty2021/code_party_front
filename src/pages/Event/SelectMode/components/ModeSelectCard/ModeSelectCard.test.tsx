import React from "react";
import { ModeSelectCard } from "./ModeSelectCard";
import "jest-styled-components";
import { render, shallow } from "enzyme";

describe("<ModeSelectCard />", () => {
  it("snapshot test", () => {
    const wrapper = render(<ModeSelectCard mode="solo" />);

    expect(wrapper).toMatchSnapshot();
  });

  it("render test black S", () => {
    expect(
      render(<ModeSelectCard mode="battle" disabled={true} />)
    ).toBeTruthy();
  });

  it("click test", () => {
    const clickHandler = jest.fn();
    const wrapper = shallow(
      <ModeSelectCard mode="solo" onClick={clickHandler} />
    );
    wrapper.simulate("click");
    expect(clickHandler).toHaveBeenCalled();
  });
});
