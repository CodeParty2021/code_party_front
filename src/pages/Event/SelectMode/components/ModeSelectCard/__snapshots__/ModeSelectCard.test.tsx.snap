// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModeSelectCard /> snapshot test 1`] = `
.c3 {
  display: inline-block;
  width: 64px;
  height: 64px;
}

.c3 > span,
.c3 > div,
.c3 > svg {
  display: block;
  width: 100%;
  height: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  width: 616px;
  height: 744px;
  position: relative;
  background: #F5F6F8;
  border-radius: 64px;
  font-weight: 700;
  color: #2F364D;
  fill: #2F364D;
}

.c1 .modeselectcard_container_image {
  height: 412px;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  position: relative;
}

.c1 .modeselectcard_container_image > img {
  height: 100%;
  width: 100%;
  object-fit: cover;
}

.c1 .modeselectcard_container_main {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 32px;
  padding: 0 16px;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
}

.c1 .modeselectcard_modename {
  font-size: 30px;
  -webkit-transform: matrix(1,0,-0.08,1,0,0);
  -ms-transform: matrix(1,0,-0.08,1,0,0);
  transform: matrix(1,0,-0.08,1,0,0);
}

.c1 .modeselectcard_description {
  font-size: 24px;
  -webkit-transform: matrix(1,0,-0.08,1,0,0);
  -ms-transform: matrix(1,0,-0.08,1,0,0);
  transform: matrix(1,0,-0.08,1,0,0);
  text-align: start;
}

.c1 .modeselectcard_border {
  position: absolute;
  width: 100%;
  height: 100%;
  border-radius: 64px;
  box-sizing: border-box;
  border: 2px solid #EAEBF1;
}

.c1:hover {
  background: #5E67F5;
  color: #FFF;
  fill: #FFF;
}

.c1:hover .modeselectcard_border {
  border: 8px solid #5E67F5;
}

.c2 {
  width: 184px;
  height: 53px;
  position: absolute;
  left: 50%;
  bottom: 0;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  fill: #F5F6F8;
}

.c0:hover .c2.c2 {
  fill: #5E67F5;
}

@supports not (gap:32px) {
  .c1 .modeselectcard_container_main > *:not(:first-child) {
    margin-top: 32px;
  }
}

<button
  class="c0 c1"
>
  <div
    class="modeselectcard_container_image"
  >
    <img
      alt="訓練モード"
      src="/img/modeselectcard_solo.png"
    />
    <svg
      class="c2"
      xlink="http://www.w3.org/1999/xlink"
      xmlns="http://www.w3.org/2000/svg"
    />
  </div>
  <div
    class="modeselectcard_container_main"
  >
    <span
      class="c3"
      display="inline-block"
    />
    <span
      class="modeselectcard_modename"
    >
      訓練モード
    </span>
    <span
      class="modeselectcard_description"
    >
      まずはこのモードでロボットを
      <br />
      動かす練習をしましょう。
    </span>
  </div>
  <div
    class="modeselectcard_border"
  />
</button>
`;
